File: frontend/src/types/embedding.ts
export type EmbeddingProviderName = 'bge' | 'e5' | 'nomic' | 'openai' | 'voyage' | 'jina';

export const EMBEDDING_PROVIDERS: { value: EmbeddingProviderName; label: string }[] = [
  { value: 'bge', label: 'BGE (local, default)' },
  { value: 'e5', label: 'E5 (local)' },
  { value: 'nomic', label: 'Nomic (local)' },
  { value: 'openai', label: 'OpenAI' },
  { value: 'voyage', label: 'Voyage' },
  { value: 'jina', label: 'Jina' },
];

export type EmbeddingVersionStatus = 'pending' | 'ready' | 'failed';
export type EmbeddingStatus = 'ready' | 'failed';

export interface EmbeddingVersion {
  id: string;
  chunk_set_id: string;
  document_id: string;
  provider: string;
  model: string;
  dimensions: number;
  version: number;
  status: EmbeddingVersionStatus;
  status_message: string | null;
  embedding_count: number;
  total_tokens: number;
  total_cost_usd: number | null;
  avg_latency_ms: number | null;
  created_at: string;
  updated_at: string;
}

export interface Embedding {
  id: string;
  embedding_version_id: string;
  chunk_id: string;
  token_count: number;
  cost_usd: number | null;
  latency_ms: number;
  status: EmbeddingStatus;
  status_message: string | null;
}

export interface EmbeddingVersionComparison {
  version_a: EmbeddingVersion;
  embeddings_a: Embedding[];
  version_b: EmbeddingVersion;
  embeddings_b: Embedding[];
}

File: backend/app/schemas/embedding.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.embedding import EmbeddingStatus, EmbeddingVersionStatus


class EmbeddingVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    chunk_set_id: uuid.UUID
    document_id: uuid.UUID
    provider: str
    model: str
    dimensions: int
    version: int
    status: EmbeddingVersionStatus
    status_message: str | None
    embedding_count: int
    total_tokens: int
    total_cost_usd: float | None
    avg_latency_ms: int | None
    created_at: datetime
    updated_at: datetime


class EmbeddingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    embedding_version_id: uuid.UUID
    chunk_id: uuid.UUID
    token_count: int
    cost_usd: float | None
    latency_ms: int
    status: EmbeddingStatus
    status_message: str | None


class GenerateEmbeddingsRequest(BaseModel):
    provider: str = Field(..., min_length=1, max_length=20)
    model: str | None = Field(default=None, max_length=100)


class EmbeddingVersionComparison(BaseModel):
    version_a: EmbeddingVersionRead
    embeddings_a: list[EmbeddingRead]
    version_b: EmbeddingVersionRead
    embeddings_b: list[EmbeddingRead]
